General updates to createResources.sh

gadams999 opened the pull request:
**This pull request enables the rest of the `createResources.sh` steps.**

Example `createResources.sh` has a mixed case `ROOT_NAME`, which is used as part of the S3 bucket creation. Per the [S3 bucket restrictions](http://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html), one of the criteria for a bucket name is *Bucket names can contain lowercase letters, numbers, and hyphens*. This will fail during the creation of the bucket.

This pull request sets the `ROOT_NAME` portion of the bucket name to all lower case. Comment also added to indicate this. 

Also corrected grep search for JSON where element order not guaranteed; specific AWS account needed for associating IAM role to Cognito Identity Pool. Captured in a later commit on my fork.

gadams999 committed:
Set bucket name to lowercase

gadams999 committed:
Update createResources.sh

quantuminformation commented on aws/createResources.sh:
```diff
@@ -1,7 +1,8 @@
 ROOT_NAME=DevDay2
-BUCKET_NAME=budilov-$ROOT_NAME
+# Bucket name must be all lowercase, and start/end with lowecase letter or number
+BUCKET_NAME=budilov-${ROOT_NAME,,}
```
${} makes lowercase?

gadams999 replied:
`${stringname,,}` (double comma's) will lowercase the $stringname variable. ^^ will uppercase.

Note: Newer versions of bash needed to support Unicode. [StackExchange question in it](http://askubuntu.com/questions/383352/command-to-convert-an-upper-case-string-to-lower-case).

gadams999 replied:
Ironic, OS X El Cap (10.11.x) only has bash version 3.x. Had to revert to another method for making the bucket name lowercase.

vbudilov replied:
$ROOT_NAME  is a variable and it would resolve to DevDay2 -- which shell are you running this script in? This same command works for me.

gadams999 committed:
Enabled pool, client app, and identity roles

For user-pool and user-pool-client id's, changed grep to greedliy capture zero to all characters after the key-value pair. Returned JSON sometimes had ID values as the last pair, hence no `,` which would return a NULL value.

The `set-identity-pool-roles` requires an AWS account number as part of the ARN. Added `AWS_ACCOUNT` variable, which should be set to the account this script is run against.

gadams999 committed:
bash v3 compatible lowercase code

Didn't realize OS X 10.11 (El Capitan) has bash 3.x instead of the more recent 4.x. The $(echo ...) code works in all versions of bash.

gadams999 committed:
Changed grep regex to OS X supported

OS X doesn't support perl style regex. Assumption is that for each returned JSON, there is only one "Id": or "ClientId":
line per file of interest. Reduced the regex to only look for that. Works in OS X.

Wonder if this file can be replaced with a CloudFormation template to do this, maybe with a Lambda helper, if needed.

vbudilov reviewed (commented):
Thanks for the updates! I'll have to check this out and run it...Probably after re:invent :)

gadams999 commented:
No worries, I'm here too. :) will update with some more context. 

Sent from my iPhone

> On Nov 28, 2016, at 16:45, Vladimir Budilov <notifications@github.com> wrote:
> 
> @vbudilov commented on this pull request.
> 
> Thanks for the updates! I'll have to check this out and run it...Probably after re:invent :)
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

vbudilov commented on aws/createResources.sh:
```diff
@@ -1,8 +1,12 @@
 
 ROOT_NAME=DevDay2
-BUCKET_NAME=budilov-$ROOT_NAME
+# Bucket name must be all lowercase, and start/end with lowecase letter or number
```
It doesn't seem to be the case. I just ran the below command and the bucket was created:
aws s3 mb s3://BUDILOVV-MYBUCKET --region us-east-1

vbudilov commented on aws/createResources.sh:
```diff
@@ -1,8 +1,12 @@
+# Bucket name must be all lowercase, and start/end with lowecase letter or number
+# $(echo...) code to work with versions of bash older than 4.0
+BUCKET_NAME=budilov-$(echo "$ROOT_NAME" | tr '[:upper:]' '[:lower:]')
 TABLE_NAME=LoginTrail$ROOT_NAME
```
Also, I'm testing this on a Mac with bash v3.2 and everything works:
ac87a3120d3c:~ budilovv$ ROOT_NAME=DevDay
ac87a3120d3c:~ budilovv$ BUCKET_NAME=budilov-$ROOT_NAME
ac87a3120d3c:~ budilovv$ echo $BUCKET_NAME
budilov-DevDay
ac87a3120d3c:~ budilovv$ bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin15)
Copyright (C) 2007 Free Software Foundation, Inc.
ac87a3120d3c:~ budilovv$

vbudilov commented on aws/createResources.sh:
```diff
@@ -43,21 +47,21 @@ cat authrole.json | sed 's/TABLE_NAME/'$TABLE_NAME'/' > /tmp/authrole.json
-#userPoolId=$(grep -Po '"Id":.*?[^\\]",' /tmp/$POOL_NAME-create-user-pool | awk -F'"' '{print $4}')
-#echo "Created user pool with an id of " $userPoolId
+aws cognito-idp create-user-pool --pool-name $POOL_NAME --auto-verified-attributes email --policies file://user-pool-policy.json --region $REGION > /tmp/$POOL_NAME-create-user-pool
+userPoolId=$(grep -E '"Id":' /tmp/$POOL_NAME-create-user-pool | awk -F'"' '{print $4}')
```
What wasn't working here? Why the switch from -Po to -E?

gadams999 replied:
On OS X, grep command doesn't support `-Po` flag (perl regex). The `-E` does work on OS X and Ubuntu (and probably others).

From what I can tell, we're only interested in the line of JSON with the `Id:` key. Did not see duplicates in there, so this should provide awk with what's needed.

vbudilov commented on aws/createResources.sh:
```diff
@@ -43,21 +47,21 @@ cat authrole.json | sed 's/TABLE_NAME/'$TABLE_NAME'/' > /tmp/authrole.json
-#echo "Created user pool with an id of " $userPoolId
+aws cognito-idp create-user-pool --pool-name $POOL_NAME --auto-verified-attributes email --policies file://user-pool-policy.json --region $REGION > /tmp/$POOL_NAME-create-user-pool
+userPoolId=$(grep -E '"Id":' /tmp/$POOL_NAME-create-user-pool | awk -F'"' '{print $4}')
+echo "Created user pool with an id of " $userPoolId
```
Also, the above command is still incomplete, from my memory at least, since the email requirement needs to be set.

gadams999 replied:
Correct, the parameter `--schema Name=email,Required=true` will check the email attribute as required. I can provide another pull or let you change on master or a new branch.

vbudilov replied:
I've added this functionality already. Thanks!
